### PR TITLE
Added support for loading webdatasets

### PR DESCRIPTION
**Pull request by djghosh13**
Should work for locally-stored webdatasets.

Files need to be named specifically as `wds_DATASET_SPLIT.tar[.gz]`, where `--dataset` should be passed in as `wds/DATASET` and `SPLIT` is specified by `--split`, and the TAR file should be located directly in the directory specified by `--dataset_root`. For zeroshot classification, the classnames should be stored in a text file in the same directory named `wds_DATASET_classes.txt`.

The webdatasets themselves need to have images with either .jpg or .png extensions, and labels with .cls extensions.

Also: Modified KITTI dataset version to 3.3.0, forgot to separate that out.

### CODE REVIEW

**Comment by rom1504** [comment]
Did you test it already? If yes, can you share the webdataset files to HF and the building script here ?

**Comment by djghosh13** [comment]
I have tested it on my end, but I'll need to upload to HF and finalize my building script.

**Comment by djghosh13** [comment]
CLIP_benchmark should work for HTTP URLs now! I've uploaded many but not all datasets to [my Huggingface account](https://huggingface.co/djghosh), so you can test it yourself without permanent downloads.

None of the previous file-naming requirements apply anymore. The following file structure is expected:
```
dataset_root/
    train/
        nshards.txt
        0.tar
        1.tar
        ...
    test/
        nshards.txt
        0.tar
        1.tar
        ...
    classnames.txt
    zeroshot_classification_templates.txt
```

This has only been tested with zeroshot classification so far, but since it only affects dataset loading, it should work for linear probing as well. For datasets stored online on Huggingface, `dataset_root` should point to the file tree, not just the repo (see example below).

Example (adapted from README):
```
clip_benchmark \
    --dataset=wds/cifar10 \
    --task=zeroshot_classification \
    --pretrained=laion400m_e32 \
    --model=ViT-B-32-quickgelu \
    --output=result.json \
    --batch_size=64 \
    --dataset_root=https://huggingface.co/datasets/djghosh/wds_cifar10_test/tree/main
```

**Comment by djghosh13** [comment]
Added a CLI command to convert a CLIP_benchmark dataset to the webdataset format. The dataset will only be stored locally; upload to Huggingface is left to the user for now.

File: clip_benchmark/datasets/builder.py [hunk 1]
```diff
@@ -535,27 +536,75 @@ def build_tfds_dataset(name, transform, download=True, split="test", data_dir="r
     return ds
 
 
-def build_wds_dataset(dataset_name, transform, download=True, split="test", data_dir="root"):
-    """Load a dataset in WebDataset format (for now only local files). Must be in a .tar or .tar.gz"""
+def build_wds_dataset(dataset_name, transform, download=False, split="test", data_dir="root"):
+    """Load a dataset in WebDataset format. Only local files for now. Expected file structure is:
```
Comment (rom1504) [inline clip_benchmark/datasets/builder.py:540]: I think only local files is wrong now ?

**Comment by rom1504** [comment]
looks great!

a few comments:
> it should work for linear probing as well

that would need the training set too but I see only tests in HF, is it expected ?

I see you didn't put imagenet1k, should it be there too ?

Also one last point: could you add a few lines in the readme to explain the new command line tool you added and a note on how to use the new wds/ feature ? (maybe even an option in https://github.com/LAION-AI/CLIP_benchmark/blob/main/benchmark/run.sh to append the wds/ prefix ?)

**Comment by rom1504** [comment]
I'll try it on my side very soon.

Do you observe the same results with wds and with original ?

**Comment by rom1504** [comment]
> 
Added a CLI command to convert a CLIP_benchmark dataset to the webdataset format. The dataset will only be stored locally; upload to Huggingface is left to the user for now.

ok, could be great to put one line of explanation about that in the readme

File: clip_benchmark/datasets/builder.py (hunk 1 above)
Comment (djghosh13) [reply clip_benchmark/datasets/builder.py:540]: You're right, I forgot to change that before

**Comment by djghosh13** [comment]
> that would need the training set too but I see only tests in HF, is it expected ?
Each split is converted separately, but will be saved in the same folder. So for example you can run

```
clip_benchmark_export_wds --dataset cifar10 --split train --dataset_root infolder/ --output outfolder/
clip_benchmark_export_wds --dataset cifar10 --split test --dataset_root infolder/ --output outfolder/
```
and you will end up with both splits stored in `outfolder/wds_cifar10/`.

I just tried running it with linear probe on cifar10; it fails specifically on getting the length of the dataloader ([here](https://github.com/LAION-AI/CLIP_benchmark/blob/607f96464cadcafaeff96c138348f555179caafb/clip_benchmark/metrics/linear_probe.py#L197)) because WebDataset doesn't have a length by default (it can be manually specified with dataset.with_length(n)).

> I see you didn't put imagenet1k, should it be there too ?

I haven't converted every dataset yet, and specifically have left out some larger ones for now; ImageNet will be there eventually.

> Also one last point: could you add a few lines in the readme to explain the new command line tool you added and a note on how to use the new wds/ feature ? (maybe even an option in https://github.com/LAION-AI/CLIP_benchmark/blob/main/benchmark/run.sh to append the wds/ prefix ?)

Will do.

**Comment by djghosh13** [comment]
> Do you observe the same results with wds and with original ?

Yes, for evaluation, the accuracies and other stats listed are exactly the same, as long as you convert to WEBP (the default) or PNG; obviously, compressing to JPEG may cause differences. However, one thing that I still haven't sorted out is that the data order may be changed even without explicit shuffling, so this may be another complication to solve before saying it works for linear probe (where training data order may matter for reproducibility).

**Comment by djghosh13** [comment]
I've updated the readme to describe how to both create and evaluate on webdatasets. I also made a change to the `clip_benchmark_export_wds` command to make the data directory consistent with what you pass in to the `clip_benchmark` command.

File: README.md [hunk 2]
```diff
@@ -118,6 +118,56 @@ Example with `cifar10`:
 
  `pip install pycocotools`
 
+ ### Webdataset example
+
+Here is an example on how to run it on [webdatasets](https://github.com/webdataset/webdataset).
+First, you need to install `webdataset`.
+
+`pip install webdataset`
```
Comment (rom1504) [inline README.md:126]: what about we simply add it as dep ?

**Comment by rom1504** [comment]
ok thanks, looks good! I'll try it myself before merging

remaining for another PR:
* run for all datasets (best if providing a script for this)
* consider making it the default in run.sh

File: README.md (hunk 2 above)
Comment (djghosh13) [reply README.md:126]: We could do that: I was mostly keeping in line with the VTAB and TFDS examples, where it's installed separately.

**Comment by gabrielilharco** [comment]
hi, quick follow up, are there any plans to merge this?

**Comment by rom1504** [comment]
Yeah
I wanted to check it but found there's a bug that prevents running main https://github.com/LAION-AI/CLIP_benchmark/issues/49

So need to fix that first

**Comment by gabrielilharco** [comment]
got it, thanks for the update!

**Comment by rom1504** [comment]
trying now.

comparison on cifar
* accuracy exactly the same
* computation time is 27s for files ; 39s for wds ; probably due to too few shards

Comments
* would be good to put all datasets at hf
* then to reference all the datasets links in a file
* and adapt run.sh to have an option to use webdataset as a source, so it works for all datasets
* important to have enough shards (at least 4)

**Comment by rom1504** [comment]
I'm thinking let's merge, and address these points in another PR

**Comment by rom1504** [comment]
thanks, and sorry for the delay!

**Comment by rom1504** [comment]
follow up tracked there https://github.com/LAION-AI/CLIP_benchmark/issues/52

**Comment by gabrielilharco** [comment]
thanks for testing and merging!
